Compatibility with the new SSA convention

fritzalder opened the pull request:
These are small changes in the linker and the sm_entry and sm_exit stubs to make the linker compatible with the upcoming SSA changes (see https://github.com/sancus-tee/sancus-core/pull/24 )

The linker contains SSA labels for:
- ssa_base which is also ssa_pc
- ssa_sp which is at ssa_base-2
- ssa_end which is at ssa_base-28

Additionally, the linker also has a ssa_thread_id which lies above the SSA at ssa_base+2 for later compatibility when adding actual threading to the SSAs. This can be removed if you find it unclean to add this now.

Also, since we technically only need ssa_base from now on, we could refactor all stubs to recalculate the SP or PC etc from ssa_base on instead of keeping the ssa_sp and ssa_pc labels. But I am not sure whether sm_pc is needed elsewhere so I refrained from renaming it.

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -14,9 +14,19 @@ __sm_entry:
+    mov #__sm_ssa_base, &__sm_ssa_base_addr
+    
+1:  
+    ; Now switch the stack to the ssa base
```
```suggestion
```

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -14,9 +14,19 @@ __sm_entry:
+    ; __sm_sp is our stackpointer, it lies at #ssa_base-2
     mov &__sm_sp, r1
     ; initialize sp on first entry
     cmp #0x0, r1
```
below it says "safe to handle IRQs now" which is *not* true afais! With the new SSA frame convention, the reti logic itself is _not_ interruptible so the SSA has to be copied in the reg file in case of an IRQ return _before_ IRQs are re-enabled...

fritzalder replied:
I moved the safe to handle IRQs to later in the sm_entry and into the reti entry. There we still need to resolve the pop r2 issue but besides that this is resolved I think

jovanbulck commented on src/stubs/sm_exit.s:
```diff
@@ -56,11 +58,14 @@ __reti_entry:
     pop r13
     pop r14
     pop r15
+    pop r2
```
careful, moving here in R2 also re-enables IRQs, but the SP and R1 have not yet been restored so might be lost right?!

fritzalder replied:
I see two options here:
1) ignore it as we plan to change the behavior of r2 anyway. Then, an eint in r2 would not have an effect anyway
2) Pop r2 into a different register, remove the gint bit and move it into r2

fritzalder replied:
btw this is resolved by now with a mix: 1) GIE bit is cleared before the pop r2 line and writes to IRQ bit are ignored by the atomicity monitor anyway if it is not the SM ID 1 running.

jovanbulck commented:
> These are small changes in the linker and the sm_entry and sm_exit stubs to make the linker compatible with the upcoming SSA changes (see [sancus-tee/sancus-core#24](https://github.com/sancus-tee/sancus-core/pull/24) )
> 
> The linker contains SSA labels for:
> 
>     * ssa_base which is also ssa_pc
> 
>     * ssa_sp which is at ssa_base-2
> 
>     * ssa_end which is at ssa_base-28
> 
> 
> Additionally, the linker also has a ssa_thread_id which lies above the SSA at ssa_base+2 for later compatibility when adding actual threading to the SSAs. This can be removed if you find it unclean to add this now.

Okay great, we can keep this for now I think. Even when having one SSA and a scheduler we might need this. Also, later we should have a field here for the caller_id.

> Also, since we technically only need ssa_base from now on, we could refactor all stubs to recalculate the SP or PC etc from ssa_base on instead of keeping the ssa_sp and ssa_pc labels. But I am not sure whether sm_pc is needed elsewhere so I refrained from renaming it.

Hm, I see you aliased sm_sp now to the SSA location. I think this is _not_ always okay though(!) The current stubs use sm_sp in the entry/exit logic for inter-enclave calls/rets. Be aware that the SSA registers are supposed to be fully transparent and can be overwritten at any moment (outside dint/eint region) by the HW IRQ logic. Any global variables like _sm_sp would have to be stored above the SSA frame, as a thread-local var.

To make this clearer, I suggest to rename things to `__sm_ssa_blah` to make it clear they're in the SSA frame?

fritzalder commented on src/stubs/sm_entry.s:
```diff
@@ -14,9 +14,19 @@ __sm_entry:
-    ; __sm_sp is our stackpointer
-    mov #__sm_sp, &__sm_sp_addr
+
+    ; initialize SSA on first entry (if ssa_thread_id is 0x0)
```
I fixed this in the latest commit and made the stack setup simpler

fritzalder commented:
Latest changes:
1) __sm_sp is moved between __sm_ssa_end and __sm_stack_init. This now gives us one static region to store our SP independent of the SSA. The old location of __sm_sp is renamed to __sm_ssa_sp. Changes in sm_entry for this are restricted to:
- When restoring SP initially, we first use SSA_SP and if it is empty, we use SP.
- To check whether an IRQ happened, we check SSA_SP
- When restoring state in reti, we read from and clear SSA_SP
- All other locations keep using SP

2) In addition to thread_id I added a caller_id field above the SSA. caller_id is filled after the check for an IRQ call and is only filled if it is empty (on the first call). Only when returning, caller_id is emptied. This means that during any intermediate call into the SM, the caller_id remains static until it returns.

3) Linker can now take a YAML configuration file. An example is in the drivers subdirectory where linker.py lies.

4) Added an alternative default sm_isr_basic stub that does not perform any stack pushes before entering the ISR. Can be enabled through the new YAML configuration file.

3) and 4) are technically not part of this PR but are reasonably the next step that we wanted to merge into master. We can split this off if you prefer.

jovanbulck commented on src/sancus_support/sm_support.h:
```diff
@@ -566,6 +572,7 @@ always_inline sm_id sancus_get_caller_id(void)
 {
```
sancus_get_caller_id should be replaced with either a compiler intrinsic or (prob easier/better) a hand-coded asm function in the sm_entry stub file to return the stored caller ID from the SSA frame right?

fritzalder replied:
Good point and this is an interesting one. We already store the caller id in the `sm_ssa_caller_id` field that is at ssa_bas + 4. However, this means that we need the SM name to get to this address. After looking at the other functions in sm_support.h, my first instinct was to add the below function:
```C
always_inline sm_id sancus_get_original_caller_id(struct SancusModule* sm)
{
    sm_id ret;
    asm("mov %1 r15\n\t"
        "mov r15 %0"
        : "=m"(ret)
        // secret_end - 2 is ssa_base_addr that points to ssa_base. 
        // We want sm_caller_id that lies at ssa_base + 4
        : "m"(*( (void **)(sm->secret_end - 2)) + 4)
        : "15"
    );
    return ret;
}
```

This would technically work with the downside that it is an additional function (and not backward compatible). However, this calculation would rely on the untrusted `SancusModule` struct and I am pretty sure that the attacker could modify that pointer temporarily to trick the SM. Interestingly, the sm_support.h function `sancus_verify` should have that same problem, correct?

I see two more options: 
1. Implement get_caller_id as a compiler define that takes the sancus name and inserts it into the `sm_sancus_##name##_caller_id` address.
2. Automatically generate a `sm_##name##_get_caller_id` with each SM_DEFINE call.

I'm not too sure which is the best option (or whether I missed one)

jovanbulck replied:
yes this is an interesting problem, I should probably have clarified a bit better what I meant above:

* compiler intrinsic:  the function is defined but not implemented here, during the compiler pass (when the name of the enclave is known), the compiler intercepts all invocations and pastes in the required asm instruction (parameterized with the enclave name). this is essentially a big ugly hack, but the developer luckily doesn't see it :)
* prob easier/better: a hand-coded asm function in the sm_entry stub file. Just declare the C function in the header file, and provide a hand-written asm implementation somewhere in the sm_entry/exit files (or provide a new file as with sm_verify.S). Then you can use a generic symbol name __sm_ssa or so, which is rewritten by our custom linker to include the enclave name (such a symbol already exists I think)

hope this makes it clearer?

jovanbulck replied:
for reference, for the compiler intrinsic, I found my old commit in an old downstream version of the Sancus compiler to do exactly this (even though I think the 2nd approach above might be somewhat neater, more maintainable): 

```diff
commit cc54aa451bd41c183218fc91211ae6d2d6e06efd
Author: Jo Van Bulck <jo.vanbulck@cs.kuleuven.be>
Date:   Thu May 26 14:56:31 2016 +0200

    Implemented sancus_get_caller_id as a compiler intrinsic.
    
    CallerID is stored atomically in SM-private TCS memory on SM entry
    to ensure it remains stable in a preemptive environment. Compiler-
    generated sancus_get_caller_id function transparently fetches the
    callerID from the current SM TCS.

diff --git a/src/llvm/SancusModuleCreator.cpp b/src/llvm/SancusModuleCreator.cpp
index 8b54108..c61c7f4 100644
--- a/src/llvm/SancusModuleCreator.cpp
+++ b/src/llvm/SancusModuleCreator.cpp
@@ -42,6 +42,7 @@ struct SancusModuleCreator : ModulePass
     Constant* getSymbolAddress(Module& m, StringRef name);
     std::string fixSymbolName(const std::string& name);
     CallSite handleSancusCall(CallSite cs);
+    CallSite handleSancusGetCallerId(CallSite cs);
 
     typedef std::map<const GlobalValue*, SancusModuleInfo> InfoMap;
     InfoMap modulesInfo;
@@ -166,6 +167,12 @@ bool SancusModuleCreator::handleFunction(Function& f)
                 handleSancusCall(cs).getInstruction();
             continue;
         }
+        else if (callee->getName() == "sancus_get_caller_id")
+        {
+            replacements[cs.getInstruction()] =
+                handleSancusGetCallerId(cs).getInstruction();
+            continue;
+        }
 
         Function* stub = getStub(f, *callee);
         cs.setCalledFunction(stub);
@@ -531,3 +538,24 @@ CallSite SancusModuleCreator::handleSancusCall(CallSite cs)
     auto asmCall = CallInst::Create(inlineAsm, args);
     return CallSite(asmCall);
 }
+
+CallSite SancusModuleCreator::handleSancusGetCallerId(CallSite cs)
+{
+    auto callerInfo = getSancusModuleInfo(cs.getCaller());
+
+    /* callerID is stored in SM Thread Control Structure, below TCS stack */
+    auto asmStr = Twine("mov &" + callerInfo.getStackPointerAddress() + ", $0\n\t"
+                        "sub #" + callerInfo.getStackSize() + "+2, $0\n\t"
+                        "mov @$0, $0"
+                        ).str();
+    auto constraintsStr = "=r";
+    auto asmFuncTy = FunctionType::get(wordTy, /*isVarArg=*/false);
+
+    // NOTE hasSideEffects has to be true because the inline assembly may be
+    // optimised away otherwise
+    auto inlineAsm = InlineAsm::get(asmFuncTy, asmStr, constraintsStr,
+                                    /*hasSideEffects=*/true);
+
+    auto asmCall = CallInst::Create(inlineAsm);
+    return CallSite(asmCall);
+}
diff --git a/src/llvm/SancusModuleInfo.cpp b/src/llvm/SancusModuleInfo.cpp
index 0ce4262..77f4177 100644
--- a/src/llvm/SancusModuleInfo.cpp
+++ b/src/llvm/SancusModuleInfo.cpp
@@ -25,6 +25,16 @@ std::string SancusModuleInfo::getExitName() const
     return "__sm_" + name + "_exit";
 }
 
+std::string SancusModuleInfo::getStackSize() const
+{
+    return "__sm_" + name + "_stack_size";
+}
+
+std::string SancusModuleInfo::getStackPointerAddress() const
+{
+    return "__sm_" + name + "_sp_addr";
+}
+
 std::string SancusModuleInfo::getIndexName(const std::string& entry) const
 {
     return "__sm_" + name + "_entry_" + entry + "_idx";
diff --git a/src/llvm/SancusModuleInfo.h b/src/llvm/SancusModuleInfo.h
index abc3c48..d1ec81b 100644
--- a/src/llvm/SancusModuleInfo.h
+++ b/src/llvm/SancusModuleInfo.h
@@ -14,6 +14,8 @@ struct SancusModuleInfo
     std::string getTableSection() const;
     std::string getEntryName() const;
     std::string getExitName() const;
+    std::string getStackSize() const;
+    std::string getStackPointerAddress() const;
     std::string getIndexName(const std::string& entry) const;
     std::string getCalleeIdName(const std::string& callee) const;
     std::string getCalleeMacName(const std::string& callee) const;
diff --git a/src/sancus_support/sm_support.h b/src/sancus_support/sm_support.h
index adc798b..e5b9988 100644
--- a/src/sancus_support/sm_support.h
+++ b/src/sancus_support/sm_support.h
@@ -283,18 +283,11 @@ always_inline sm_id sancus_get_self_id(void)
  *
  * The calling module is defined as the previously executing module. That is,
  * the module that entered the currently executing module.
+ *
+ * @note This function is implemented as a compiler intrinsic to ensure the
+ * callerID remains stable in a preemptive environment.
  */
-always_inline sm_id sancus_get_caller_id(void)
-{
-    //TODO use __sm_sp_addr and an offset to retrieve the callerID stored on entry
-    sm_id ret;
-    asm(".word 0x1387\n\t"
-        "mov r15, %0"
-        : "=m"(ret)
-        :
-        : "r15");
-    return ret;
-}
+sm_id sancus_get_caller_id(void);
 
 /**
  * Perform a call to a Sancus module.
```

jovanbulck replied:
Update re caller_id behavior changes:

- [x] added tests in sancus-examples to make sure this behavior will work as expected (see commit referenced above from sancus-examples repo). This will fail after merging and some assertions will have to be updated to the new behavior.
- [ ] we need to also provide a version of sancus_get_caller_id when called by unprotected code? This may be simply the instruction itself, or `unprotected_entry.S` could be updated to store the caller_id somewhere? Not sure how to select the correct function at link time -- maybe the compiler intrinsic would be easier in this respect?
- [ ] backwards compatibility will be partially broken with the new behavior; this is fine, but we need to update the exact definition of this function in the sancus_support doc above. You can prob start from your description above:
> caller_id field above the SSA. caller_id is filled after the check for an IRQ call and is only filled if it is empty (on the first call). Only when returning, caller_id is emptied. This means that during any intermediate call into the SM, the caller_id remains static until it returns.

jovanbulck replied:
also, not sure, but why don't we make life easier and *do* retain backwards compatibility with a non-preemptive environment by simply overwriting the ssa_caller_id field on every SM entry or re-entry from an ocall (ie only when the call is valid and will be accepted of course)?

This would be compatible with the current definition and would not break any existing applications afais?

Arguments against this proposed behavior?

fritzalder replied:
I'm not completely sure what you mean with the last point, to overwrite the ssa_caller_id on every call. That is basically already what happens. See https://github.com/sancus-tee/sancus-compiler/pull/33/files#diff-b4f90c1290867b2f6af0d36a10035e6735c84fb624c69d6230f75aa081dab112R67 Or do you want to overwrite the HW caller ID field based on the SSA entry? That would not be the simplest solution I'd say.

My argument about breaking backwards compatibility was that we will need to include the name of the SM into the function call (either into the name or as an argument). And that would break bw compatibility. I do not think that using a compiler intrinsic is the best approach as it would completely hide away the actual function that is used from the developer. I think adding an inline `sm_##name##_get_original_caller_id` function when defining the SM would be a simple and effective solution that also forces the developer to think about what they're calling.

I can still go for the compiler intrinsic version similar to your above code if you prefer. Should be an easy commit since we don't need the additional calculations since we have the sm_##name##_caller_id address already available.

jovanbulck replied:
no, we don't want to have the developer using another function. Hiding this from the developer in a compiler intrinsic is a desirable solution I'd say :)

what I meant w backwards compatibility is purely about the semantics of the returned ID: I am a bit confused w your description above:

> This means that during any intermediate call into the SM, the caller_id remains static until it returns.

So, the question is: is the returned ID (1) the one of the previously executing SM in this thread (ie as would be when no scheduler and IRQs would be involved), or (2) is the ID the one of the *original* caller of this ECALL and not updated on OCALLs as I was under the impression from your desciption? My point is we want (1) not (2), but not sure what we have atm?

To make it specific: will the tests I added in sancus-examples continue to work or will they fail (as I thought when adding the TODO lines)?

https://github.com/sancus-tee/sancus-examples/blob/master/arithmetic/foo.c#L19

jovanbulck replied:
from a quick look, it seems like it's behavior (1) above, so that's good?

but the entry stub is currently very hard to read w many flows and Im not sure Aion covers all of them atm? Let's try to clarify at least in comments the atomic periods: how long is the HW entry atomic period, from where to where is it extended with clix, how does __sm_isr handling fit in the clix period etc?

Re the latter, can/should we make a stub without ISR handlign for Aion?

jovanbulck replied:
also,re changes to sm_entry/exit, linking 2 outstanding severe bug issues that ideally should be fixed here as well: #13 and #22

jovanbulck replied:
especially atomicity requirements for linking 2 SMs with sancus_verify as outlined in #22 are relevant to the Aion design and not yet properly handled in the current implementation afaik :thinking:

jovanbulck replied:
after discussion with @fritzalder : local attestation and jumping to another SM entry point only needs to happen atomically *if* the callee may unload itself using `sancus_disable` *without* notifying potential callers first.

We concluded this is not a concern to be transparently fixed by hardware/runtime at the moment, i.e., secure linking is currently *not* meant to be interrupt-safe. This PR should add a short note for the application developer in the documentation of `sancus_disable` to clarify that SMs should not call this function when they might still be called by a trusted caller SM (and the caller SM can verify this is indeed not the case by local attestation of the callee SM code section).

fritzalder replied:
I added a warning to sancus_disable and some comments to the entry stub. Feel free to do another iteration over it.

jovanbulck replied:
done, clarified a bit further sancus_disable comment

jovanbulck replied:
compiler intrinsic for `sancus_get_caller_id` looks good! 

I realize we only need a solution still for `sancus_verify_caller`. However, we should drop this primitive as it can easily be implemented with other primitives as follows: `sancus_verify(adrs) + sancus_get_id(adrs) == sancus_get_caller_id()`

jovanbulck replied:
disabled in sancus_support.h and left a note as per 33a1d1a 

In principle we can transparently handle this case with support from the sm_entry enclave runtime, but I don't think anyone actually uses sancus_verify_caller(). If so, we can always do this later.

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -7,35 +7,42 @@
+    mov r15, &__sm_tmp
+    ; set up clix length and call clix (word 0x1389)
+    mov #100, r15 
+    .word 0x1389
```
not sure anymore: how will `clix` behave when ATOMICITY_MONITOR=0 in sancus-core? It should be a NOP right?

jovanbulck replied:
also a bit tricky to have 100 hard-coded here maybe? let's at least indicate with a big line in comments from where to where it runs atomically?

fritzalder replied:
Yes
- `ATOMICITY_MONITOR=0` : `clix` is a `NOP and `dint` disables interrupts
- `ATOMICITY_MONITOR=1` : `clix` disables interrupts for X cylces and `dint` is a `NOP`

fritzalder replied:
Done in latest commit

jovanbulck reviewed (commented):
partial review of compiler

jovanbulck commented:
I added some notes on  caller_id behavior changes above.

Thinking further about it, I think you *do* need to implement this with a compiler intrinsic, at least partially. Otherwise you'd have to pass the name of the calling SM all the time to rewrite the symbol to the function that will be inserted by the linker.

Maybe the cleanest/most maintainable is a hybrid approach:

1. use a compiler intrinsic as in the code above to intercept all calls to `sancus_get_caller_id()`.
2. in the compiler pass, rewrite these calls to a unique symbol name like `__sm_NAME_get_caller_id()` or `__unprotected_get_caller_id()` (the latter may also be the direct `.word` machine instruction, depending on the behavior we want for unprotected code)
3. in the linker, rename the generic implementation of `__sm_get_caller_id()` to the unique `__sm_NAME_get_caller_id()`

jovanbulck commented on src/stubs/sm_exit.s:
```diff
@@ -80,6 +80,14 @@ __reti_entry:
     .global __ret_entry
     .type __ret_entry,@function
 __ret_entry:
+    ; If we return from an OCALL, overwrite the stored caller_id
```
hm, not sure, but I think this may be insecure(!)

The atomic entry perion `sm_entry.S` already finished here so doing a callerID instruction does *not* guarantee this is still the prev SM in the thread right?

Why don't we reserve a variable to store the callerID in the atomic entry and then we can move it optionally into the SSA later?

fritzalder replied:
Good catch. 
Alternatively we move the eint to after the branch and add an eint in the ret entry after the pop r15 line. I am not sure what is arguably more overhead, adding a variable for this or extending the atomic section by some few cycles? I would vote for extending the atomic section.

jovanbulck replied:
no, I vote for keep the atomic section as short as possible :)

and also it's much cleaner and easier to understand when the atomic section does not cross files I find, a bit unfortunate that's already not the case for sm_reti_entry and sm_isr though

jovanbulck replied:
okay I think I fixed it as per 7b51d8e

@fritzalder  pls double-check (there was also an incorrect # instead of & I think) and mark resolved when done ^^

jovanbulck replied:
to be secure, the sm_entry stub should still verify that there's actually an open entry point when accepting a call with 0xffff

this is described in issue #13 

let's discuss this next week

fritzalder replied:
This part looks good and I'll resolve it. Further discussion on #13 will be done there to not clutter up this PR discussion even more :D

jovanbulck commented on src/sancus_support/sm_support.h:
```diff
@@ -580,6 +586,20 @@ always_inline sm_id sancus_get_caller_id(void)
     return ret;
 }
 
+// always_inline sm_id sancus_get_original_caller_id(struct SancusModule* sm)
```
not sure if this is for future, but let's remove this better altogether, cause the `struct SancusModule` is typically stored and managed in *untrusted* code, so not to be trusted

fritzalder replied:
No, sorry that was my local draft that made it into the commit when I blindly added the sancus_disable comments. It's already gone in my code and will be removed in the next commit.

fritzalder commented:
I added a commit that implements the compiler intrinsic for `get_caller_id`. This is my first time working with Twine so I am not sure that this is indeed correct. Especially the `$0` in the assembly that I added are added without any basis, they just looked right after looking at the other code. This works on my machine but I will need to revisit this on Monday to see whether the generated assembly code really makes sense.

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -55,9 +63,29 @@ __sm_entry:
+    ; We are not called by an IRQ. If __sm_ssa_caller_id is not set, fill it with the caller id
+    mov #__sm_ssa_caller_id, r15
+    cmp #0x0, r15
+    jne 1f
```
```suggestion
    jeq 1f
    cmp #0xffff, r6
    jne 2f
1:
```

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -55,9 +63,29 @@ __sm_entry:
+    .word 0x1387
+    mov r15, &__sm_ssa_caller_id
+
+1:
```
```suggestion
2:
```

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -55,9 +63,29 @@ __sm_entry:
+
 1:
+    ; We are not called by an IRQ. If __sm_ssa_caller_id is not set, fill it with the caller id
+    mov #__sm_ssa_caller_id, r15
```
```suggestion
    tst &__sm_ssa_caller_id
```

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -55,9 +63,29 @@ __sm_entry:
 1:
+    ; We are not called by an IRQ. If __sm_ssa_caller_id is not set, fill it with the caller id
+    mov #__sm_ssa_caller_id, r15
+    cmp #0x0, r15
```
```suggestion
```

jovanbulck commented on src/stubs/sm_entry.s:
```diff
@@ -55,9 +63,29 @@ __sm_entry:
     br #__sm_isr
+
 1:
+    ; We are not called by an IRQ. If __sm_ssa_caller_id is not set, fill it with the caller id
```
```suggestion
    ; We are not called by an IRQ. Fill __sm_ssa_caller_id on ecall and on ocall return.
```

jovanbulck commented:
> I added a commit that implements the compiler intrinsic for `get_caller_id`. This is my first time working with Twine so I am not sure that this is indeed correct. Especially the `$0` in the assembly that I added are added without any basis, they just looked right after looking at the other code. This works on my machine but I will need to revisit this on Monday to see whether the generated assembly code really makes sense.

Not very experienced myself about working with Twine, but looks good! :+1: I only removed the line "mov r15, $0" cause I think that will become "mov r15, r15"

jovanbulck commented:
I'd like #13 to be still fixed before merging, because it directly affects the availability guarantees of Aion and should be fixable now that we're already rewriting the entry/exit stubs thoroughly.

Best to have this all merged at once, so we can update the CI tests (and nemesis/sancus-step calibration etc) as needed and hopefully have everything working nicely together :)

fritzalder commented:
#13 is now addressed in the newest commits. 
1. I had to revert the SancusModuleCreator to still do the mov operation after caller_id since the compiler does use r12, not r15 for returns. This then gave an error in the hello-world example as it expected the caller_id in r12.
2. The earlier added check for r6=0xffff is not enough as it allows the attacker to use this actively to overwrite the ssa_caller_id field. I now explicitly also check for ocall_id. I argue that checking for 0xffff in r6 is still useful to make the reentry explicit, but arguably we do not need it anymore now that we check for ocall_id. I would leave it in.
3. We have some code duplication now for the double checking of OCALL reentry. But I did not see a nice way to unify these 10 lines and anything that I came up with was more than 10 lines.

jovanbulck commented on src/stubs/sm_exit.s:
```diff
@@ -56,11 +66,24 @@ __reti_entry:
 
-    reti
+    ; branch to pc (r2 already restored manually above)
+    br &__sm_pc
```
not sure if calling/returning SMs then works with the current compiler/core?

jovanbulck commented:
wait now I'm confused myself :thinking: 

does one SSA mean that nested ecalls are not allowed? where is this enforced in the stub?

jovanbulck commented:
no .Lerror is a local label

and `error` is a stdlib function outside the enclave (hence my comments regarding leaking state and rewriting .Lerror to an infinite loop, not the cleanest but secure as opposed to what we had previously)

I'm now preparing a commit proposal fix
